ambry-replication/src/main/java/com.github.ambry.replication/CloudToStoreReplicationManager.java

```diff
@@ -0,0 +1,333 @@
+  }
+
+  @Override
+  public void start() throws ReplicationException {
+    // Add listener for vcr instance config changes
+    vcrClusterSpectator.registerInstanceConfigChangeListener(new InstanceConfigChangeListener() {
+      @Override
+      public void onInstanceConfigChange(List<InstanceConfig> instanceConfigs, NotificationContext context) {
+        ConcurrentSkipListSet<CloudDataNode> newVcrNodes = new ConcurrentSkipListSet<>();
+        ConcurrentHashMap<String, CloudDataNode> newInstanceNameToCloudDataNode = new ConcurrentHashMap<>();
+        Set<CloudDataNode> removedNodes = new HashSet<>(vcrNodes);
+        synchronized (notificationLock) {
+          for (InstanceConfig instanceConfig : instanceConfigs) {
+            String instanceName = instanceConfig.getInstanceName();
+            Port sslPort =
+                getSslPortStr(instanceConfig) == null ? null : new Port(getSslPortStr(instanceConfig), PortType.SSL);
+            CloudDataNode cloudDataNode = new CloudDataNode(instanceConfig.getHostName(),
+                new Port(Integer.valueOf(instanceConfig.getPort()), PortType.PLAINTEXT), sslPort,
+                clusterMapConfig.clustermapVcrDatacenterName, clusterMapConfig);
+            newInstanceNameToCloudDataNode.put(instanceName, cloudDataNode);
+            newVcrNodes.add(cloudDataNode);
+          }
+          removedNodes.removeAll(newVcrNodes);
+          vcrNodes = newVcrNodes;
+          instanceNameToCloudDataNode = newInstanceNameToCloudDataNode;
+          List<PartitionId> partitionsOnRemovedNodes = getPartitionsOnRemovedNodes(removedNodes);
+          for (PartitionId partitionId : partitionsOnRemovedNodes) {
+            try {
+              removeCloudReplica(partitionId.toPathString());
+              addCloudReplica(partitionId.toPathString());
+            } catch (ReplicationException rex) {
+              logger.error("Could not remove/add replica for partitionId {}", partitionId);
+            }
+          }
+        }
+      }
+    });
+
+    // Add listener for vcr instance liveness changes
+    vcrClusterSpectator.registerLiveInstanceChangeListener(new LiveInstanceChangeListener() {
+      @Override
+      public void onLiveInstanceChange(List<LiveInstance> liveInstances, NotificationContext changeContext) {
+        ConcurrentSkipListSet<CloudDataNode> newVcrNodes = new ConcurrentSkipListSet<>();
+        synchronized (notificationLock) {
+          for (LiveInstance liveInstance : liveInstances) {
+            if (instanceNameToCloudDataNode.containsKey(liveInstance.getInstanceName())) {
+              newVcrNodes.add(instanceNameToCloudDataNode.get(liveInstance.getInstanceName()));
+            }
+          }
+          Set<CloudDataNode> removedNodes = new HashSet<>(vcrNodes);
+          removedNodes.removeAll(newVcrNodes);
+          vcrNodes = newVcrNodes;
+          List<PartitionId> partitionsOnRemovedNodes = getPartitionsOnRemovedNodes(removedNodes);
```

Comment:
This section looks identical or very similar to one in previous method.  Consolidate?